models/ou_noise_generator.cpp

```diff
@@ -0,0 +1,377 @@
+  , logger_( n )
+{
+}
+
+/* ----------------------------------------------------------------
+ * Parameter extraction and manipulation functions
+ * ---------------------------------------------------------------- */
+
+void
+nest::ou_noise_generator::Parameters_::get( DictionaryDatum& d ) const
+{
+  ( *d )[ names::mean ] = mean_;
+  ( *d )[ names::std ] = std_;
+  ( *d )[ names::dt ] = dt_.get_ms();
+  ( *d )[ names::tau ] = tau_;
+}
+
+void
+nest::ou_noise_generator::State_::get( DictionaryDatum& d ) const
+{
+}
+
+void
+nest::ou_noise_generator::Parameters_::set( const DictionaryDatum& d, const ou_noise_generator& n, Node* node )
+{
+  updateValueParam< double >( d, names::mean, mean_, node );
+  updateValueParam< double >( d, names::std, std_, node );
+  updateValueParam< double >( d, names::tau, tau_, node );
+  if ( tau_ <= 0 )
+  {
+    throw BadProperty( "tau > 0 required." );
+  }
+  double dt;
+  if ( updateValueParam< double >( d, names::dt, dt, node ) )
+  {
+    dt_ = Time::ms( dt );
+  }
+  if ( std_ < 0 )
+  {
+    throw BadProperty( "The standard deviation cannot be negative." );
+  }
+
+  if ( not dt_.is_step() )
+  {
+    throw StepMultipleRequired( n.get_name(), names::dt, dt_ );
+  }
+}
+
+
+/* ----------------------------------------------------------------
+ * Default and copy constructor for node
+ * ---------------------------------------------------------------- */
+
+nest::ou_noise_generator::ou_noise_generator()
+  : StimulationDevice()
+  , P_()
+  , S_()
+  , B_( *this )
+{
+  recordablesMap_.create();
+}
+
+nest::ou_noise_generator::ou_noise_generator( const ou_noise_generator& n )
+  : StimulationDevice( n )
+  , P_( n.P_ )
+  , S_( n.S_ )
+  , B_( n.B_, *this )
+{
+}
+
+
+/* ----------------------------------------------------------------
+ * Node initialization functions
+ * ---------------------------------------------------------------- */
+
+void
+nest::ou_noise_generator::init_state_()
+{
+  StimulationDevice::init_state();
+}
+
+void
+nest::ou_noise_generator::init_buffers_()
+{
+  StimulationDevice::init_buffers();
+  B_.logger_.reset();
+
+  B_.next_step_ = 0;
+  B_.amps_.clear();
+  B_.amps_.resize( P_.num_targets_, 0.0 );
+}
+
+void
+nest::ou_noise_generator::pre_run_hook()
+{
+  B_.logger_.init();
+
+  StimulationDevice::pre_run_hook();
+  if ( P_.num_targets_ != B_.amps_.size() )
+  {
+    LOG( M_INFO, "ou_noise_generator::pre_run_hook()", "The number of targets has changed, drawing new amplitudes." );
+    init_buffers_();
+  }
+
+  V_.dt_steps_ = P_.dt_.get_steps();
+
+  const double h = Time::get_resolution().get_ms();
+  const double t = kernel().simulation_manager.get_time().get_ms();
+
+  // scale Hz to ms
+  const double noise_amp = P_.std_ * std::sqrt( -1 * std::expm1( -2 * h / P_.tau_ ) );
+  const double prop = std::exp( -1 * h / P_.tau_ );
+  const double tau_inv = -std::expm1( -h / P_.tau_ );
+
+  V_.noise_amp_ = noise_amp;
+  V_.prop_ = prop;
+  V_.tau_inv_ = tau_inv;
```

Comment:
I would simplify this a bit

```suggestion
  V_.noise_amp_ = P_.std_ * std::sqrt( -1 * std::expm1( -2 * h / P_.tau_ ) );
  V_.prop_ = std::exp( -1 * h / P_.tau_ );
  V_.tau_inv_ = -std::expm1( -h / P_.tau_ );
```